lib/search/fragmentation_scheme.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Search.FragmentationScheme do
+  @doc """
+  Splits a binary into multiple binaries that satisfy limitations specified by opts.
+
+  If possible, splits the text on whitespace to preserve words. If that is impossible, splits text in between graphemes.
+
+  Supported options:
+
+  * `:max_size` - maximum byte_size of the output binaries. The output binaries may have size less or equal to that
+    value, which also should guarantee the sequence length after tokenization will be bounded by this value.
+  """
+  def split(text, opts \\ [])
+  def split("", _opts), do: []
+
+  def split(text, opts) when is_binary(text) do
+    case Keyword.get(opts, :max_size) do
+      nil ->
+        [text]
+
+      max_size ->
+        text
+        |> compute_splits({0, 0}, 0, max_size, [])
+        |> split_binary(0, text, [])
+    end
+  end
+
+  def recombine(chunks), do: Enum.join(chunks)
+
+  defp split_binary([], _start_idx, _binary, acc), do: Enum.reverse(acc)
+
+  defp split_binary([split_size | splits_tail], start_idx, binary, acc) do
+    current_part = binary_slice(binary, start_idx, split_size)
+    split_binary(splits_tail, start_idx + split_size, binary, [current_part | acc])
+  end
+
+  defp compute_splits("", {0, 0}, 0, _max_size, acc), do: Enum.reverse(acc)
+
+  defp compute_splits("", {chunk_size, trailing_whitespace}, next_word, max_size, acc) do
+    if chunk_size + trailing_whitespace + next_word <= max_size do
+      compute_splits("", {0, 0}, 0, max_size, [chunk_size + trailing_whitespace + next_word | acc])
+    else
+      compute_splits("", {0, 0}, 0, max_size, [next_word, chunk_size + trailing_whitespace | acc])
+    end
+  end
+
+  defp compute_splits(
+         text,
+         {current_chunk_size, trailing_whitespace_size},
+         next_word_size,
+         max_size,
+         acc
+       ) do
+    {graph, text} = String.next_grapheme(text)
+    graph_size = byte_size(graph)
+    whole_chunk_size = current_chunk_size + trailing_whitespace_size
+
+    if graph =~ ~r/\s/ do
+      # graph is whitespace
+      if next_word_size == 0 do
+        # we are still building the current chunk
+        if whole_chunk_size + graph_size <= max_size do
```

Comment:
I have a feeling we could reduce the number of conditionals, but I may be missing something. What I have in mind is this:

```
if chunk_size + grapheme_size > max_chunk_size do
  if "the current chunk has no finished word (in other words, no whitespace)" do
    # split at the current point (mid-word) and emit chunk
  else
    # emit chunk until the last seen word ending
  end
else
  # if whitespace, record new word ending
end
```

The high-level idea is that we keep track of the current position, the current chunk start position and a "possible" chunk ending. We bump the "possible" chunk ending whenever we get a whitespace. Once we overflow the max size we need to emit a chunk, which is either:

* if "possible" chunk ending > chunk start, that's our chunk
* otherwise it means there was no whitespace, so the chunk is from chunk start all the way to the current position

Reply:
I just thought about it agin and realised I was unnecessarily thinking in terms of offsets, so this can be even simpler:

```elixir
do_split(string, max_size, 0, nil, [])

defp do_split("", _, size, _, sizes), do: Enum.reverse(sizes, [size])

defp do_split(string, max_size, size, size_until_word, sizes) do
  {grapheme, string} = String.next_grapheme(string)
  grapheme_size = byte_size(grapheme)

  if size + grapheme_size > max_size do
    if size_until_word do
      # Split before the current unfinished word
      next = size - size_until_word
      do_split(string, max_size, next + grapheme_size, nil, [size_until_word | sizes])
    else
      # The current chunk has a single word, just split it
      do_split(string, max_size, grapheme_size, nil, [size | sizes])
    end
  else
    new_size = size + grapheme_size
    size_until_word = if whitespace?(grapheme), do: new_size, else: size_until_word
    do_split(string, max_size, new_size, size_until_word, sizes)
  end
end
```

Where `size_until_word` is the chunk size until (including) the last whitespace, or `nil` if there was no whitespace.

I think this is close to your implementation, just with compressed the conditionals :)

Reply:
Sorry for the late reply, had to take a day off. You are right, the function does the same thing in 1/3 the code haha. The way I was looking at it made me convinced the problem was a lot more complex than it actually was... The wonders of code review